// draft-js-side-toolbar-plugin/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _decorateComponentWithProps = require('decorate-component-with-props');

var _decorateComponentWithProps2 = _interopRequireDefault(_decorateComponentWithProps);

var _createStore = require('./utils/createStore');

var _createStore2 = _interopRequireDefault(_createStore);

var _Toolbar = require('./components/Toolbar');

var _Toolbar2 = _interopRequireDefault(_Toolbar);

var _DefaultBlockTypeSelect = require('./components/DefaultBlockTypeSelect');

var _DefaultBlockTypeSelect2 = _interopRequireDefault(_DefaultBlockTypeSelect);

var _buttonStyles = {
  "buttonWrapper": "draftJsToolbar__buttonWrapper__1Dmqh",
  "button": "draftJsToolbar__button__qi1gf",
  "active": "draftJsToolbar__active__3qcpF",
  "separator": "draftJsToolbar__separator__3M3L7"
};

var _buttonStyles2 = _interopRequireDefault(_buttonStyles);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var createSideToolbarPlugin = function createSideToolbarPlugin() {
  var config = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  var store = (0, _createStore2.default)({
    isVisible: false
  });

  var _config$structure = config.structure;
  var structure = _config$structure === undefined ? [_DefaultBlockTypeSelect2.default] : _config$structure;


  var toolbarProps = {
    store: store,
    structure: structure,
    buttonTheme: _buttonStyles2.default
  };

  return {
    initialize: function initialize(_ref) {
      var setEditorState = _ref.setEditorState;
      var getEditorState = _ref.getEditorState;
      var getEditorRef = _ref.getEditorRef;

      store.updateItem('getEditorState', getEditorState);
      store.updateItem('setEditorState', setEditorState);
      store.updateItem('getEditorRef', getEditorRef);
    },
    // Re-Render the toolbar on every change
    onChange: function onChange(editorState) {
      store.updateItem('editorState', editorState);
      return editorState;
    },
    SideToolbar: (0, _decorateComponentWithProps2.default)(_Toolbar2.default, toolbarProps)
  };
};

exports.default = createSideToolbarPlugin;
