// draft-js-mathjax-plugin/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _draftJs = require('draft-js');

var _utils = require('./utils');

var _insertTeX2 = require('./modifiers/insertTeX');

var _insertTeX3 = _interopRequireDefault(_insertTeX2);

var _InlineTeX = require('./components/InlineTeX');

var _InlineTeX2 = _interopRequireDefault(_InlineTeX);

var _TeXBlock = require('./components/TeXBlock');

var _TeXBlock2 = _interopRequireDefault(_TeXBlock);

var _loadMathJax = require('./loadMathJax');

var _loadMathJax2 = _interopRequireDefault(_loadMathJax);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var createMathjaxPlugin = function createMathjaxPlugin(_ref /* config */) {
  var macros = _ref.macros;

  (0, _loadMathJax2.default)({ macros: macros });

  var store = {
    getEditorState: undefined,
    setEditorState: undefined,
    getReadOnly: undefined,
    setReadOnly: undefined,
    getEditorRef: undefined,
    teXToUpdate: {}
  };

  var _insertTeX = function _insertTeX() {
    var block = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;

    var editorState = store.getEditorState();
    store.setEditorState((0, _insertTeX3.default)(editorState, block));
  };

  var insertChar = function insertChar(char) {
    var editorState = store.getEditorState();
    var sel = editorState.getSelection();
    var offset = sel.getStartOffset() - 1;
    var newContentState = _draftJs.Modifier.replaceText(editorState.getCurrentContent(), sel.merge({
      anchorOffset: offset,
      focusOffset: offset + 1
    }), char);
    store.setEditorState(_draftJs.EditorState.push(editorState, newContentState, 'insert-characters'));
  };

  var keyBindingFn = function keyBindingFn(e, _ref2) {
    var getEditorState = _ref2.getEditorState;
    return (0, _utils.myKeyBindingFn)(getEditorState)(e);
  };

  var blockRendererFn = function blockRendererFn(block) {
    if (block.getType() === 'atomic' && block.getData().get('mathjax')) {
      return {
        component: _TeXBlock2.default,
        editable: false,
        props: { getStore: function getStore() {
            return store;
          } }
      };
    }
    return null;
  };

  // l'utilisation des flèches gauche ou droite
  // amène le curseur sur une formule
  var updateTeX = function updateTeX(key, dir) {
    // le composant associé à la formule
    // se sert de cet indicateur
    // pour se reconnaître
    // cf.componentWillReceiveProps
    store.teXToUpdate = { key: key, dir: dir };
    var editorState = store.getEditorState();
    store.setEditorState(_draftJs.EditorState.forceSelection(editorState, editorState.getSelection()));
  };

  var handleKeyCommand = function handleKeyCommand(command /* ,{ getEditorState, setEditorState } */) {
    if (command === 'insert-texblock') {
      _insertTeX(true);
      return 'handled';
    }
    if (command === 'insert-inlinetex') {
      _insertTeX();
      return 'handled';
    }
    // command de la forme 'enter-inline-math-<dir>-<entityKey>',
    // lancée lorsque l'utilisateur déplace le curseur
    // sur une formule à l'aide des flèches gauche/droite(dir:l ou r)
    if (command.slice(0, 16) === 'update-inlinetex') {
      var dir = command.slice(17, 18);
      var entityKey = command.slice(19);
      updateTeX(entityKey, dir);
      return 'handled';
    }
    if (command.slice(0, 15) === 'update-texblock') {
      var _dir = command.slice(16, 17);
      var blockKey = command.slice(18);

      updateTeX(blockKey, _dir);
      return 'handled';
    }
    if (command.slice(0, 11) === 'insert-char') {
      var char = command.slice(12);
      insertChar(char);
      return 'handled';
    }
    return 'not-handled';
  };

  return {
    initialize: function initialize(_ref3) {
      var getEditorState = _ref3.getEditorState,
          setEditorState = _ref3.setEditorState,
          getReadOnly = _ref3.getReadOnly,
          setReadOnly = _ref3.setReadOnly,
          getEditorRef = _ref3.getEditorRef;

      store.getEditorState = getEditorState;
      store.setEditorState = setEditorState;
      store.getReadOnly = getReadOnly;
      store.setReadOnly = setReadOnly;
      store.getEditorRef = getEditorRef;
    },
    decorators: [{
      strategy: _utils.findInlineTeXEntities,
      component: _InlineTeX2.default,
      props: { getStore: function getStore() {
          return store;
        } }
    }],
    keyBindingFn: keyBindingFn,
    handleKeyCommand: handleKeyCommand,
    blockRendererFn: blockRendererFn
  };
};

exports.default = createMathjaxPlugin;
